File: user-service/src/main/java/com/ilogos/user/jwt/JwtService.java
package com.ilogos.user.jwt;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.KeyFactory;
import java.security.interfaces.RSAPublicKey;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;
import java.util.Optional;
import java.util.function.Consumer;

import org.springframework.http.HttpStatus;
import org.springframework.security.oauth2.jwt.JwtDecoder;
import org.springframework.stereotype.Service;

import com.ilogos.shared.model.TokenInfo;
import com.ilogos.shared.utils.TokenInfoUtils;
import com.ilogos.user.exception.ExceptionWithStatus;

import jakarta.annotation.PostConstruct;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@RequiredArgsConstructor
@Service
public class JwtService {

    private final JwtConfig jwtConfig;

    private RSAPublicKey publicKey;

    private RSAPublicKey loadKey() throws IOException {
        String pem = Files.readString(Path.of(jwtConfig.getPublicKeyPath()));

        if (pem == null || pem.isBlank()) {
            throw new IllegalStateException("Public key is missing");
        }
        String key = pem.replace("-----BEGIN PUBLIC KEY-----", "")
                .replace("-----END PUBLIC KEY-----", "")
                .replaceAll("\\s", "");
        byte[] decoded = Base64.getDecoder().decode(key);
        X509EncodedKeySpec keySpec = new X509EncodedKeySpec(decoded);
        try {
            KeyFactory factory = KeyFactory.getInstance("RSA");
            return (RSAPublicKey) factory.generatePublic(keySpec);
        } catch (Exception e) {
            throw new RuntimeException("Unable to load RSA public key", e);
        }
    }

    @PostConstruct
    public void init() throws IOException {
        if (this.jwtConfig != null) {
            publicKey = loadKey();
        }
    }

    public TokenInfo getTokenInfo(String token) {
        return TokenInfoUtils.INSTANCE.createInfo(token, publicKey);
    }

    public TokenInfo extractTokenInfoFromHeader(String header) {
        var token = header.startsWith("Bearer ")
                ? Optional.of(header.substring(7))
                : Optional.<String>empty();
        if (token.isEmpty()) {
            log.info("Bearer token not setted");
            throw new ExceptionWithStatus(HttpStatus.UNAUTHORIZED, "Bearer token not setted");
        }
        return TokenInfoUtils.INSTANCE.createInfo(token.get(), publicKey);
    }

    public JwtDecoder buildJwtDecoder() {
        Consumer<String> logger = (String msg) -> log.info(msg);
        return TokenInfoUtils.INSTANCE.buildJwtDecoder(publicKey, logger);
    }

}

File: course-service/src/main/kotlin/com/ilogos/courseManager/jwt/JwtService.kt
package com.ilogos.courseManager.jwt

import com.ilogos.courseManager.exception.ExceptionWithStatus
import com.ilogos.shared.model.TokenInfo
import com.ilogos.shared.utils.TokenInfoUtils
import jakarta.annotation.PostConstruct
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Profile
import org.springframework.http.HttpStatus
import org.springframework.security.oauth2.jwt.JwtDecoder
import org.springframework.stereotype.Service
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.security.KeyFactory
import java.security.interfaces.RSAPublicKey
import java.security.spec.X509EncodedKeySpec
import java.util.*
import java.util.function.Consumer


@Service
class JwtService(private val _jwtConfig: JwtConfig?) {

    private lateinit var publicKey: RSAPublicKey

    val jwtConfig: JwtConfig
        get() = _jwtConfig ?: throw IllegalStateException("Jwt-service has not been initialized")

    companion object {
        private val log = LoggerFactory.getLogger(JwtService::class.java)

        @Profile("test")
        fun create(publicKey: RSAPublicKey): JwtService {
            val service = JwtService(null)
            service.publicKey = publicKey

            return service
        }
    }

    @Throws(IOException::class)
    private fun loadKey(): RSAPublicKey {
        val pem = Files.readString(Path.of(jwtConfig.publicKeyPath))
        if (pem.isBlank()) throw IllegalStateException("Private key is missing")

        val key = pem.replace("-----BEGIN PUBLIC KEY-----", "").replace("-----END PUBLIC KEY-----", "")
            .replace("\\s".toRegex(), "")

        val decoded = Base64.getDecoder().decode(key)
        val keySpec = X509EncodedKeySpec(decoded)
        return try {
            KeyFactory.getInstance("RSA").generatePublic(keySpec) as RSAPublicKey
        } catch (e: Exception) {
            throw RuntimeException("Unable to load RSA key", e)
        }
    }

    @PostConstruct
    @Throws(IOException::class)
    fun init() {
        if (_jwtConfig != null) {
            val key = loadKey()
            publicKey = key
        }
    }

    fun getTokenInfo(token: String): TokenInfo = TokenInfoUtils.createInfo(token, publicKey)

    fun extractTokenInfoFromHeader(header: String): TokenInfo {
        val token = if (header.startsWith("Bearer ")) header.removePrefix("Bearer ").trim() else ""
        if (token.isBlank()) {
            log.info("Bearer token not set")
            throw ExceptionWithStatus(HttpStatus.UNAUTHORIZED, "Bearer token not set")
        }

        return TokenInfoUtils.createInfo(token, publicKey)
    }

    fun buildJwtDecoder(): JwtDecoder {
        val logger: Consumer<String> = Consumer { msg: String -> log.info(msg) }
        return TokenInfoUtils.buildJwtDecoder(publicKey, logger)
    }
}
